File: ios/Tests/FlowbizOnsiteTests/FixtureSupport.swift
import Foundation
@testable import FlowbizOnsite

/// Test-side helpers for the shared drift-guard fixtures (`shared/fixtures/`):
/// locating the fixture directory, mapping fixture `input` JSON onto the typed
/// constructors, and structural JSON comparison.
enum FixtureSupport {

    struct FixtureError: Error, CustomStringConvertible {
        let description: String
        init(_ description: String) { self.description = description }
    }

    /// `shared/<name>` resolved relative to this source file
    /// (`ios/Tests/FlowbizOnsiteTests/` → repo root → `shared/<name>`).
    static func sharedDirectory(_ name: String) -> URL {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent() // FixtureSupport.swift
            .deletingLastPathComponent() // FlowbizOnsiteTests
            .deletingLastPathComponent() // Tests
            .deletingLastPathComponent() // ios
            .appendingPathComponent("shared/\(name)", isDirectory: true)
    }

    static func fixturesDirectory() -> URL {
        sharedDirectory("fixtures")
    }

    static func fixtureFiles() throws -> [URL] {
        let directory = fixturesDirectory()
        let contents = try FileManager.default.contentsOfDirectory(
            at: directory, includingPropertiesForKeys: nil
        )
        return contents
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
    }

    static func loadFixture(_ url: URL) throws -> [String: Any] {
        let data = try Data(contentsOf: url)
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw FixtureError("\(url.lastPathComponent): root is not an object")
        }
        return object
    }

    /// Maps a fixture (`event` name + camelCase `input`) onto the typed constructors.
    static func buildEvent(_ eventName: String, input: [String: Any]) throws -> Event {
        switch eventName {
        case "pageView":
            return .pageView(path: input["path"] as? String, title: input["title"] as? String)
        case "accountLogin":
            return .accountLogin(user: try user(try object(input, "user")))
        case "accountSync":
            return .accountSync(user: try user(try object(input, "user")))
        case "productView":
            return .productView(product: try product(try object(input, "product")))
        case "cartSync":
            return .cartSync(cart: try cart(try object(input, "cart")))
        case "addToCart":
            return .addToCart(products: try objectArray(input, "products").map(cartItem))
        case "cartItemUpdate":
            return .cartItemUpdate(
                cartId: try string(input, "cartId"),
                productId: try string(input, "productId"),
                sku: try string(input, "sku"),
                quantity: try int(input, "quantity")
            )
        case "cartSetPostalCode":
            return .cartSetPostalCode(
                cartId: try string(input, "cartId"),
                postalCode: try string(input, "postalCode")
            )
        case "cartSetCoupon":
            return .cartSetCoupon(
                cartId: try string(input, "cartId"),
                coupon: try string(input, "coupon")
            )
        case "checkoutStep":
            let json = try object(input, "checkout")
            return .checkoutStep(checkout: Checkout(
                cartId: try string(json, "cartId"),
                step: try int(json, "step"),
                totalSteps: try int(json, "totalSteps"),
                stepName: try string(json, "stepName")
            ))
        case "orderComplete":
            return .orderComplete(order: try order(try object(input, "order")))
        case "orderCancel":
            return .orderCancel(
                orderId: input["orderId"] as? String,
                cartId: input["cartId"] as? String
            )
        default:
            throw FixtureError("unknown fixture event: \(eventName)")
        }
    }

    private static func user(_ json: [String: Any]) throws -> User {
        User(
            userId: try string(json, "userId"),
            email: try string(json, "email"),
            phone: json["phone"] as? String,
            name: json["name"] as? String,
            plan: json["plan"] as? String,
            createdAt: json["createdAt"] as? String
        )
    }

    private static func product(_ json: [String: Any]) throws -> Product {
        Product(
            productId: try string(json, "productId"),
            url: json["url"] as? String,
            category: json["category"] as? String,
            brand: json["brand"] as? String,
            variants: try objectArray(json, "variants").map(variant)
        )
    }

    private static func variant(_ json: [String: Any]) throws -> ProductVariant {
        ProductVariant(
            sku: try string(json, "sku"),
            price: try double(json, "price"),
            name: json["name"] as? String,
            url: json["url"] as? String,
            imageUrl: json["imageUrl"] as? String,
            priceFrom: doubleOrNil(json, "priceFrom"),
            stock: intOrNil(json, "stock"),
            available: boolOrNil(json, "available"),
            properties: try propertiesOrNil(json, "properties"),
            recoveryProperties: try propertiesOrNil(json, "recoveryProperties")
        )
    }

    private static func cartItem(_ json: [String: Any]) throws -> CartItem {
        CartItem(
            productId: try string(json, "productId"),
            sku: try string(json, "sku"),
            quantity: try int(json, "quantity"),
            price: try double(json, "price"),
            name: json["name"] as? String,
            priceFrom: doubleOrNil(json, "priceFrom"),
            category: json["category"] as? String,
            brand: json["brand"] as? String,
            url: json["url"] as? String,
            imageUrl: json["imageUrl"] as? String,
            properties: try propertiesOrNil(json, "properties"),
            recoveryProperties: try propertiesOrNil(json, "recoveryProperties")
        )
    }

    private static func address(_ json: [String: Any]) -> Address {
        Address(
            postalCode: json["postalCode"] as? String,
            addressLine1: json["addressLine1"] as? String,
            addressNumber: json["addressNumber"] as? String,
            addressLine2: json["addressLine2"] as? String,
            city: json["city"] as? String,
            state: json["state"] as? String,
            country: json["country"] as? String,
            neighborhood: json["neighborhood"] as? String
        )
    }

    private static func cart(_ json: [String: Any]) throws -> Cart {
        Cart(
            cartId: try string(json, "cartId"),
            subtotal: try double(json, "subtotal"),
            total: try double(json, "total"),
            freight: try double(json, "freight"),
            tax: try double(json, "tax"),
            discounts: try double(json, "discounts"),
            currency: json["currency"] as? String,
            coupons: json["coupons"] as? [String],
            items: try (json["items"] as? [[String: Any]]).map { try $0.map(cartItem) },
            deliveryAddress: (json["deliveryAddress"] as? [String: Any]).map(address)
        )
    }

    private static func order(_ json: [String: Any]) throws -> Order {
        Order(
            cartId: try string(json, "cartId"),
            orderId: json["orderId"] as? String,
            subtotal: try double(json, "subtotal"),
            total: try double(json, "total"),
            freight: try double(json, "freight"),
            tax: try double(json, "tax"),
            discounts: try double(json, "discounts"),
            currency: json["currency"] as? String,
            coupons: json["coupons"] as? [String],
            items: try (json["items"] as? [[String: Any]]).map { try $0.map(cartItem) },
            deliveryAddress: (json["deliveryAddress"] as? [String: Any]).map(address),
            paymentMethods: try (json["paymentMethods"] as? [[String: Any]]).map { methods in
                try methods.map { PaymentMethod(type: try string($0, "type"), amount: try double($0, "amount")) }
            },
            deliveryMethods: try (json["deliveryMethods"] as? [[String: Any]]).map { methods in
                try methods.map { DeliveryMethod(type: try string($0, "type"), amount: try double($0, "amount")) }
            }
        )
    }

    // MARK: - JSON → JSONValue

    private static func propertiesOrNil(_ json: [String: Any], _ key: String) throws -> [String: JSONValue]? {
        guard let raw = json[key] as? [String: Any] else { return nil }
        var result = [String: JSONValue]()
        for (name, value) in raw {
            result[name] = try jsonValue(value)
        }
        return result
    }

    private static func jsonValue(_ any: Any) throws -> JSONValue {
        switch any {
        case let string as String:
            return .string(string)
        case let number as NSNumber:
            return isBoolean(number) ? .bool(number.boolValue) : .number(number.doubleValue)
        case is NSNull:
            return .null
        case let array as [Any]:
            return .array(try array.map(jsonValue))
        case let object as [String: Any]:
            var entries = [String: JSONValue]()
            for (key, value) in object { entries[key] = try jsonValue(value) }
            return .object(entries)
        default:
            throw FixtureError("unsupported JSON value: \(any)")
        }
    }

    private static func isBoolean(_ number: NSNumber) -> Bool {
        CFGetTypeID(number) == CFBooleanGetTypeID()
    }

    // MARK: - Structural comparison

    /// Structural comparison — key order irrelevant, numbers compared by
    /// double value (`0` == `0.0`). Returns a description of the first
    /// difference, or nil when equivalent.
    static func diff(expected: Any, actual: Any, path: String) -> String? {
        switch (expected, actual) {
        case (is NSNull, is NSNull):
            return nil
        case (let exp as [String: Any], let act as [String: Any]):
            let expKeys = Set(exp.keys)
            let actKeys = Set(act.keys)
            guard expKeys == actKeys else {
                let missing = expKeys.subtracting(actKeys).sorted()
                let extra = actKeys.subtracting(expKeys).sorted()
                return "\(path): key mismatch (missing=\(missing), unexpected=\(extra))"
            }
            for key in expKeys.sorted() {
                if let difference = diff(expected: exp[key]!, actual: act[key]!, path: "\(path).\(key)") {
                    return difference
                }
            }
            return nil
        case (let exp as [Any], let act as [Any]):
            guard exp.count == act.count else {
                return "\(path): array length \(exp.count) != \(act.count)"
            }
            for (index, pair) in zip(exp, act).enumerated() {
                if let difference = diff(expected: pair.0, actual: pair.1, path: "\(path)[\(index)]") {
                    return difference
                }
            }
            return nil
        case (let exp as NSNumber, let act as NSNumber):
            if isBoolean(exp) != isBoolean(act) {
                return "\(path): boolean/number type mismatch (expected \(exp), was \(act))"
            }
            if isBoolean(exp) {
                return exp.boolValue == act.boolValue ? nil : "\(path): expected \(exp) but was \(act)"
            }
            return exp.doubleValue == act.doubleValue ? nil : "\(path): expected \(exp) but was \(act)"
        case (let exp as String, let act as String):
            return exp == act ? nil : "\(path): expected '\(exp)' but was '\(act)'"
        default:
            return "\(path): type mismatch (expected \(type(of: expected)), was \(type(of: actual)))"
        }
    }

    // MARK: - Extraction helpers

    private static func object(_ json: [String: Any], _ key: String) throws -> [String: Any] {
        guard let value = json[key] as? [String: Any] else {
            throw FixtureError("missing object '\(key)'")
        }
        return value
    }

    private static func objectArray(_ json: [String: Any], _ key: String) throws -> [[String: Any]] {
        guard let value = json[key] as? [[String: Any]] else {
            throw FixtureError("missing object array '\(key)'")
        }
        return value
    }

    private static func string(_ json: [String: Any], _ key: String) throws -> String {
        guard let value = json[key] as? String else {
            throw FixtureError("missing string '\(key)'")
        }
        return value
    }

    private static func double(_ json: [String: Any], _ key: String) throws -> Double {
        guard let value = json[key] as? NSNumber, !isBoolean(value) else {
            throw FixtureError("missing number '\(key)'")
        }
        return value.doubleValue
    }

    private static func int(_ json: [String: Any], _ key: String) throws -> Int {
        guard let value = json[key] as? NSNumber, !isBoolean(value) else {
            throw FixtureError("missing number '\(key)'")
        }
        return value.intValue
    }

    private static func doubleOrNil(_ json: [String: Any], _ key: String) -> Double? {
        guard let value = json[key] as? NSNumber, !isBoolean(value) else { return nil }
        return value.doubleValue
    }

    private static func intOrNil(_ json: [String: Any], _ key: String) -> Int? {
        guard let value = json[key] as? NSNumber, !isBoolean(value) else { return nil }
        return value.intValue
    }

    private static func boolOrNil(_ json: [String: Any], _ key: String) -> Bool? {
        guard let value = json[key] as? NSNumber, isBoolean(value) else { return nil }
        return value.boolValue
    }
}

File: android/sdk/src/test/kotlin/br/com/flowbiz/onsite/FixtureSupport.kt
package br.com.flowbiz.onsite

import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Test-side helpers for the shared drift-guard fixtures (`shared/fixtures/`):
 * locating the fixture directory, mapping fixture `input` JSON onto the typed
 * constructors, and structural JSON comparison.
 */
object FixtureSupport {

    /** Walks up from the working directory until `shared/<name>` is found. */
    fun sharedDir(name: String): File {
        var dir: File? = File(System.getProperty("user.dir")!!).absoluteFile
        while (dir != null) {
            val candidate = File(dir, "shared/$name")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile
        }
        error("shared/$name not found above ${System.getProperty("user.dir")!!}")
    }

    fun fixturesDir(): File = sharedDir("fixtures")

    fun fixtureFiles(): List<File> =
        fixturesDir().listFiles { f -> f.extension == "json" }!!.sortedBy { it.name }

    /** Maps a fixture (`event` name + camelCase `input`) onto the typed constructors. */
    fun buildEvent(eventName: String, input: JSONObject): Event = when (eventName) {
        "pageView" -> Event.PageView(path = input.stringOrNull("path"), title = input.stringOrNull("title"))
        "accountLogin" -> Event.AccountLogin(user(input.getJSONObject("user")))
        "accountSync" -> Event.AccountSync(user(input.getJSONObject("user")))
        "productView" -> Event.ProductView(product(input.getJSONObject("product")))
        "cartSync" -> Event.CartSync(cart(input.getJSONObject("cart")))
        "addToCart" -> Event.AddToCart(input.getJSONArray("products").objects().map { cartItem(it) })
        "cartItemUpdate" -> Event.CartItemUpdate(
            cartId = input.getString("cartId"),
            productId = input.getString("productId"),
            sku = input.getString("sku"),
            quantity = input.getInt("quantity"),
        )
        "cartSetPostalCode" -> Event.CartSetPostalCode(
            cartId = input.getString("cartId"),
            postalCode = input.getString("postalCode"),
        )
        "cartSetCoupon" -> Event.CartSetCoupon(
            cartId = input.getString("cartId"),
            coupon = input.getString("coupon"),
        )
        "checkoutStep" -> Event.CheckoutStep(
            input.getJSONObject("checkout").let {
                Checkout(
                    cartId = it.getString("cartId"),
                    step = it.getInt("step"),
                    totalSteps = it.getInt("totalSteps"),
                    stepName = it.getString("stepName"),
                )
            }
        )
        "orderComplete" -> Event.OrderComplete(order(input.getJSONObject("order")))
        "orderCancel" -> Event.OrderCancel(
            orderId = input.stringOrNull("orderId"),
            cartId = input.stringOrNull("cartId"),
        )
        else -> error("unknown fixture event: $eventName")
    }

    private fun user(json: JSONObject) = User(
        userId = json.getString("userId"),
        email = json.getString("email"),
        phone = json.stringOrNull("phone"),
        name = json.stringOrNull("name"),
        plan = json.stringOrNull("plan"),
        createdAt = json.stringOrNull("createdAt"),
    )

    private fun product(json: JSONObject) = Product(
        productId = json.getString("productId"),
        url = json.stringOrNull("url"),
        category = json.stringOrNull("category"),
        brand = json.stringOrNull("brand"),
        variants = json.getJSONArray("variants").objects().map { variant(it) },
    )

    private fun variant(json: JSONObject) = ProductVariant(
        sku = json.getString("sku"),
        price = json.getDouble("price"),
        name = json.stringOrNull("name"),
        url = json.stringOrNull("url"),
        imageUrl = json.stringOrNull("imageUrl"),
        priceFrom = json.doubleOrNull("priceFrom"),
        stock = json.intOrNull("stock"),
        available = json.boolOrNull("available"),
        properties = json.mapOrNull("properties"),
        recoveryProperties = json.mapOrNull("recoveryProperties"),
    )

    private fun cartItem(json: JSONObject) = CartItem(
        productId = json.getString("productId"),
        sku = json.getString("sku"),
        quantity = json.getInt("quantity"),
        price = json.getDouble("price"),
        name = json.stringOrNull("name"),
        priceFrom = json.doubleOrNull("priceFrom"),
        category = json.stringOrNull("category"),
        brand = json.stringOrNull("brand"),
        url = json.stringOrNull("url"),
        imageUrl = json.stringOrNull("imageUrl"),
        properties = json.mapOrNull("properties"),
        recoveryProperties = json.mapOrNull("recoveryProperties"),
    )

    private fun address(json: JSONObject) = Address(
        postalCode = json.stringOrNull("postalCode"),
        addressLine1 = json.stringOrNull("addressLine1"),
        addressNumber = json.stringOrNull("addressNumber"),
        addressLine2 = json.stringOrNull("addressLine2"),
        city = json.stringOrNull("city"),
        state = json.stringOrNull("state"),
        country = json.stringOrNull("country"),
        neighborhood = json.stringOrNull("neighborhood"),
    )

    private fun cart(json: JSONObject) = Cart(
        cartId = json.getString("cartId"),
        subtotal = json.getDouble("subtotal"),
        total = json.getDouble("total"),
        freight = json.getDouble("freight"),
        tax = json.getDouble("tax"),
        discounts = json.getDouble("discounts"),
        currency = json.stringOrNull("currency"),
        coupons = json.optJSONArray("coupons")?.let { arr -> (0 until arr.length()).map { arr.getString(it) } },
        items = json.optJSONArray("items")?.objects()?.map { cartItem(it) },
        deliveryAddress = json.optJSONObject("deliveryAddress")?.let { address(it) },
    )

    private fun order(json: JSONObject) = Order(
        cartId = json.getString("cartId"),
        orderId = json.stringOrNull("orderId"),
        subtotal = json.getDouble("subtotal"),
        total = json.getDouble("total"),
        freight = json.getDouble("freight"),
        tax = json.getDouble("tax"),
        discounts = json.getDouble("discounts"),
        currency = json.stringOrNull("currency"),
        coupons = json.optJSONArray("coupons")?.let { arr -> (0 until arr.length()).map { arr.getString(it) } },
        items = json.optJSONArray("items")?.objects()?.map { cartItem(it) },
        deliveryAddress = json.optJSONObject("deliveryAddress")?.let { address(it) },
        paymentMethods = json.optJSONArray("paymentMethods")?.objects()?.map {
            PaymentMethod(type = it.getString("type"), amount = it.getDouble("amount"))
        },
        deliveryMethods = json.optJSONArray("deliveryMethods")?.objects()?.map {
            DeliveryMethod(type = it.getString("type"), amount = it.getDouble("amount"))
        },
    )

    /**
     * Structural comparison — key order irrelevant, numbers compared by double
     * value (`0` == `0.0`). Returns a description of the first difference, or
     * null when equivalent.
     */
    fun diff(expected: Any?, actual: Any?, path: String): String? {
        val exp = if (expected == JSONObject.NULL) null else expected
        val act = if (actual == JSONObject.NULL) null else actual
        return when {
            exp == null && act == null -> null
            exp == null || act == null -> "$path: expected $exp but was $act"
            exp is JSONObject && act is JSONObject -> {
                val expKeys = exp.keyNames()
                val actKeys = act.keyNames()
                if (expKeys != actKeys) {
                    val missing = expKeys - actKeys
                    val extra = actKeys - expKeys
                    "$path: key mismatch (missing=$missing, unexpected=$extra)"
                } else {
                    expKeys.firstNotNullOfOrNull { key: String -> diff(exp.get(key), act.get(key), "$path.$key") }
                }
            }
            exp is JSONArray && act is JSONArray -> {
                if (exp.length() != act.length()) {
                    "$path: array length ${exp.length()} != ${act.length()}"
                } else {
                    (0 until exp.length()).firstNotNullOfOrNull { i -> diff(exp.get(i), act.get(i), "$path[$i]") }
                }
            }
            exp is Boolean || act is Boolean ->
                if (exp == act) null else "$path: expected $exp but was $act"
            exp is Number && act is Number ->
                if (exp.toDouble() == act.toDouble()) null else "$path: expected $exp but was $act"
            else ->
                if (exp == act) null else "$path: expected ${exp::class.simpleName}($exp) but was ${act::class.simpleName}($act)"
        }
    }

    // --- JSON extraction helpers (org.json opt* return sentinel defaults, we want nulls) ---

    private fun JSONObject.stringOrNull(key: String): String? =
        if (has(key) && !isNull(key)) getString(key) else null

    private fun JSONObject.doubleOrNull(key: String): Double? =
        if (has(key) && !isNull(key)) getDouble(key) else null

    private fun JSONObject.intOrNull(key: String): Int? =
        if (has(key) && !isNull(key)) getInt(key) else null

    private fun JSONObject.boolOrNull(key: String): Boolean? =
        if (has(key) && !isNull(key)) getBoolean(key) else null

    private fun JSONObject.mapOrNull(key: String): Map<String, Any?>? =
        optJSONObject(key)?.toPlainMap()

    private fun JSONArray.objects(): List<JSONObject> =
        (0 until length()).map { getJSONObject(it) }

    /** Android's org.json has no `keySet()`; `keys()` exists on both implementations. */
    private fun JSONObject.keyNames(): Set<String> {
        val result = linkedSetOf<String>()
        val iterator = keys()
        while (iterator.hasNext()) result += iterator.next() as String
        return result
    }

    private fun JSONObject.toPlainMap(): Map<String, Any?> =
        keyNames().associateWith { key -> get(key).toPlainValue() }

    private fun Any?.toPlainValue(): Any? = when (this) {
        null, JSONObject.NULL -> null
        is JSONObject -> toPlainMap()
        is JSONArray -> (0 until length()).map { get(it).toPlainValue() }
        else -> this
    }
}
